WalletWasabi/WabiSabi/Models/Decomposition/PossibleDecompositions.cs

```diff
@@ -0,0 +1,176 @@
+
+		private Money MaximumTotalValue { get; }
+
+		private Money MinimumTotalValue { get; }
+
+		// The final public API should only allow 2, later 3 access patterns
+		// efficiently:
+		// - When optimizing the decomposition of a specific balance, enumerate
+		//   from a specific effective cost downwards up to a limit.
+		//   Decompositions are then re-ordered after evaluating the cost
+		//   function. Below 4.0 btc maxSize can be 5 with negligible losses.
+		//   Above this amount 6 or more outputs may be required, and a
+		//   non-standard value can be forced, but generally such input balances
+		//   should just be avoided.
+		// - When evaluating likely denominations given others prevouts and
+		//   ownership proofs, we start enumerate downwards from multiple effective
+		//   values, one for each input or sum/difference of a pair of inputs,
+		//   and tally the denominations used by the closest decompositions.
+		//   Max effective cost = 2* largest input, can use round parameters to
+		//   precompute before learning prevout amounts. Max count can be hard
+		//   coded to 4 to ensure all clients can compute the same
+		//   decompositions. Likelyhood should take loss into account.
+		// - When evaluating input combinations, we need to query potentially
+		//   very large number of different balances, looking for small
+		//   decompositions with minimal losses (tight bounds). maxcount = 4? 5?
+		public IEnumerable<Decomposition> GetByTotalValue(
+			Money? maximumEffectiveCost = null,
+			Money? minimumTotalValue = null,
+			Money? minimumValue = null, // dust threshold
+			int maxOutputs = int.MaxValue,
+			int maxDecompositions = 1000,
+			FeeRate? feeRate = null,
+			int vsizePerOutput = Constants.P2WPKHOutputSizeInBytes)
+		{
+			Money costPerOutput = (feeRate ?? FeeRate.Zero).GetFee(vsizePerOutput).Satoshi;
+			var maxDecompositionCost = StratifiedDecompositions.Length * costPerOutput;
+
+			maxOutputs = Math.Max(maxOutputs, StratifiedDecompositions.Length);
+			maximumEffectiveCost ??= new Money(MaximumTotalValue);
+			Guard.True(nameof(maximumEffectiveCost), maximumEffectiveCost - maxDecompositionCost <= MaximumTotalValue, "must not exceed limit from precomputation.");
+
+			minimumTotalValue ??= MinimumTotalValue;
+			Guard.True(nameof(minimumTotalValue), MinimumTotalValue <= minimumTotalValue && minimumTotalValue <= MaximumTotalValue, "must not be lower than limit from precomputation.");
+
+			minimumValue ??= Money.Zero;
+
+			return StratifiedDecompositions
+				.Take(maxOutputs).Select((x, i) => (Decompositions: x, TotalCost: (i + 1) * costPerOutput))
+				.Select(p => p.Decompositions.Prune(
+							Money.Min(maximumEffectiveCost - p.TotalCost, MaximumTotalValue),
+							Money.Max(minimumTotalValue, MinimumTotalValue),
+							minimumValue)
+						.Where(d => d.Outputs[^1] >= minimumValue))
+				.Aggregate(MergeDescending)
+				.Take(maxDecompositions);
+		}
```

Comment:
`SoloCoinJoinTestAsync` tests fail because thete `maximumEffecticeCost - TotalCost` is less than `minimumTotalValue` by a very small value. I tried with the code below but for some reason it cannot find any decomposition with that:

```c#
`SoloCoinJoinTestAsync` test fails because the 

```c#
return StratifiedDecompositions
	.Take(maxOutputs)
	.Select((x, i) => (Decompositions: x, TotalCost: (i + 1) * costPerOutput))
	.Select(x => (x.Decompositions, x.TotalCost, MaximumTotalValue: Money.Min(maximumEffectiveCost - x.TotalCost, MaximumTotalValue), MinimumTotalValue: Money.Max(minimumTotalValue, MinimumTotalValue)))
	.Where(x => x.MaximumTotalValue - x.MinimumTotalValue > x.TotalCost)
	.Select(p => p.Decompositions.Prune(
			p.MaximumTotalValue,
			p.MinimumTotalValue,
			minimumValue)
		.Where(d => d.Outputs[^1] >= minimumValue))
	.Aggregate(MergeDescending)
	.Take(maxDecompositions);
```
```

Reply:
that means the nominal value range (specified for the precomputation) cannot accomodate for the additional cost at the given feerate